### PR TITLE
feat: tabs and historical by day

### LINKED ISSUE #13: Feature: Add a calendar widget to the UI

### LINKED ISSUE #14: Feature: Add historical data

### PR DESCRIPTION
**Pull request by gracescharf**
## Description

* Creates the Tabs component and the HistoricalByDay component.
* `<Tabs>` expects multiple React components as children and a corresponding array of labels for the table button
* `<HBD>` functions very similarly to SummaryStats -- it has a corresponding api that expects the ticker symbol and the a date (string). It makes use of `DatePicker` (a external react component) and `dayjs` (a date library similar to momentjs, but smaller) 
* The dates are limited to only monday-friday, and dates that are in the past, because the api is looking up stock price data, which isn't available on weekends or in the future 
* I haven't re-added the styles because they're on Ignacio's local... and it didn't seem like necessary learning opportunity   
* still need to improve the styling of the calendar input...

👉  You will need to run npm install once this branch is merged

## Related Issues

- Closes #13 
- Closes #14 

## Testing

1. Open build
2. Search & select a stock ticker
3. Click between summary and historical and see them switch 
4. Change the day in the calendar & see the data update
5. Change the stock ticker & see the data update

### CODE REVIEW

File: components/SummaryStats/SummaryStats.module.scss [hunk 1]
```diff
@@ -19,14 +19,14 @@
...
-    color: white;
-  }
+  //caption {
+  //   font-weight: bold;
+  //   text-transform: uppercase;
+  //   line-height: 2;
+  //   font-size: 1.3rem;
+  //   background-color: #0c2074;
+  //   color: white;
+  //}
```
Comment (gracescharf) [inline components/SummaryStats/SummaryStats.module.scss]:
```suggestion
```

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
+import React, { useEffect, useState } from "react"
+import DatePicker from "react-datepicker"
```
Comment (iprado78) [inline components/HistoricalByDay/HistoricalByDay.tsx:2]: Re-order import to follow @iprado78 's conventions

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import dayjs from "dayjs"
```
Comment (iprado78) [inline components/HistoricalByDay/HistoricalByDay.tsx:3]: I'm more familiar with `date-fns` and it is tree-shakeable.  This seems to have a smaller default size, but if we're going to introduce this library prepare to have arguments for why it fits our use cases and is better.

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+const [useHistoricalByDay] = bind(
+  combineLatest([tickerSelections$, dateSelections$]).pipe(
+    filter(([ticker, date]) => ticker !== undefined && date !== undefined),
+    switchMap(([ticker, date]) =>
+      ajax
+        .getJSON<IHistoricalByDay>(`/api/historical-by-day/${ticker}/${date}`)
+        .pipe(
+          map((historicalByDay) => {
+            console.log("historicalbyday map", historicalByDay)
```
Comment (iprado78) [inline components/HistoricalByDay/HistoricalByDay.tsx]: remove console.log

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                  label,
+                  value: formatter?.(histValue) ?? histValue,
+                }
+              },
+            )
+          }),
+        ),
+    ),
+    catchError((error) => {
+      console.error(error)
```
Comment (iprado78) [inline components/HistoricalByDay/HistoricalByDay.tsx]: it does not spark joy

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  const [date, setDate] = useState(yesterday)
+
+  const isPastWeekday = (date: Date) => {
+    const day = date.getDay()
+    return day !== 0 && day !== 6 && dayjs(date).isBefore(dayjs(), "day")
+  }
+
+  const historicalByDay = useHistoricalByDay()
+
+  console.log("historicalbyday", historicalByDay)
```
Comment (iprado78) [inline components/HistoricalByDay/HistoricalByDay.tsx]: slay

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  ),
+  [],
+)
+
+export default function HistoricalByDay() {
+  const yesterday = new Date()
+  yesterday.setDate(yesterday.getDate() - 1)
+  const [date, setDate] = useState(yesterday)
+
+  const isPastWeekday = (date: Date) => {
```
Comment (iprado78) [inline components/HistoricalByDay/HistoricalByDay.tsx]: Lift any declared/non-anonymous function that does not rely on closure out of the component, so you don't define on every render.

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    return day !== 0 && day !== 6 && dayjs(date).isBefore(dayjs(), "day")
+  }
+
+  const historicalByDay = useHistoricalByDay()
+
+  console.log("historicalbyday", historicalByDay)
+
+  useEffect(() => {
+    let formatted = dayjs(date).format("YYYYMMDD")
+    onDateSelection(formatted)
```
Comment (iprado78) [inline components/HistoricalByDay/HistoricalByDay.tsx]: onChange should handle this.  Everything else can be lifted into a stream.  Your source stream would pump out events with the unformatted date.  Then your stream (the one you export) would transform that unformatted date into the formatted one that your API needs.

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      return of(error)
+    }),
+  ),
+  [],
+)
+
+export default function HistoricalByDay() {
+  const yesterday = new Date()
+  yesterday.setDate(yesterday.getDate() - 1)
+  const [date, setDate] = useState(yesterday)
```
Comment (iprado78) [inline components/HistoricalByDay/HistoricalByDay.tsx]: We can lift all this state into a stream (see comments below).

const [useSelectedDate] = bind(dateSelectrions$.pipe(...))

File: components/StatsTable/StatsTable.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    stats: ${stats}
+  `)
+  return (
+    <div className={styles.wrapper}>
+      <table className={styles.table}>
+        <caption className="visually-hidden">{caption}</caption>
+        <tbody>
+          {stats?.map(({ label, value }) => (
+            <Row key={label} label={label} value={value} />
+          )) || null}
```
Comment (iprado78) [inline components/StatsTable/StatsTable.tsx]: null is actually redundant, I think.  if stats is undefined, then expression will return undefined and React will render empty content, same as null

File: components/SummaryStats/SummaryStats.tsx [hunk 11]
```diff
@@ -1,32 +1,29 @@
 import React from "react"
-import { map, switchMap } from "rxjs/operators"
+import { map, switchMap, tap } from "rxjs/operators"
```
Comment (iprado78) [inline components/SummaryStats/SummaryStats.tsx]: using tap?

File: components/SummaryStats/SummaryStats.tsx [hunk 12]
```diff
@@ -1,32 +1,29 @@
...
 import { tickerSelections$ } from "../TickerSearch"
+// import StatsTable from
 import {
   summaryStatsConfig,
   SummaryStatsConfigKey,
 } from "./SummaryStats.config"
 import styles from "./SummaryStats.module.scss"
+import StatsTable from "../StatsTable"
 
+// move somewhere that this can be shared
```
Comment (iprado78) [inline components/SummaryStats/SummaryStats.tsx]: components folder should have it's own models folder

File: components/Tabs/Tabs.tsx [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+  const generateButtons = () => {
+    return tabLabels.map((label, i) => {
+      let id = `${label}-${i}`
+      return (
+        <button
+          key={id}
+          role="tab"
+          className={`${styles.button} ${
+            openTab === label ? styles.isActive : ""
```
Comment (iprado78) [inline components/Tabs/Tabs.tsx]: https://www.npmjs.com/package/classnames

File: components/Tabs/Tabs.tsx [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+export default function Tabs({ tabLabels, children }: ITabsProps) {
+  // is this the right way to type react children?
+
+  const [openTab, setOpenTab] = useState(tabLabels[0])
+
+  const generateButtons = () => {
+    return tabLabels.map((label, i) => {
+      let id = `${label}-${i}`
```
Comment (iprado78) [inline components/Tabs/Tabs.tsx]: Lift into its own Button component

File: components/Tabs/Tabs.tsx [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        >
+          {label}
+        </button>
+      )
+    })
+  }
+
+  const generateTabs = () => {
+    return children.map((child, i) => {
+      let label = tabLabels[i]
```
Comment (iprado78) [inline components/Tabs/Tabs.tsx]: Lift into its own Tab component

File: package.json [hunk 16]
```diff
@@ -33,9 +33,12 @@
   "dependencies": {
     "@react-rxjs/core": "^0.6.6",
     "@react-rxjs/utils": "^0.7.0",
+    "@types/react-datepicker": "^3.1.8",
```
Comment (iprado78) [inline package.json]: Put types in dev dependencies

File: pages/api/summary-stats/[symbol].ts [hunk 17]
```diff
@@ -10,7 +10,7 @@ const fetcher = async (symbol: string) => {
     const { data } = await axios.get<ISummaryStats>(
       `${BASE_URL}/stable/stock/${symbol}/stats?token=${TOKEN}`,
     )
-    console.log("Backend ****", data)
+    // console.log("Backend ****", data)
```
Comment (iprado78) [inline pages/api/summary-stats/[symbol].ts]: Log or kill comment

File: pages/index.tsx [hunk 18]
```diff
@@ -8,9 +8,13 @@ import SummaryStats from "@/components/SummaryStats"
 import Footer from "@/components/Footer"
 import TickerSearch from "@/components/TickerSearch"
 import CompanyOverview from "@/components/CompanyOverview"
+import HistoricalByDay from "@/components/HistoricalByDay"
+import Tabs from "@/components/Tabs"
 
 const appTitle = "US Bank - Stock Analytics"
 
+// todo: hide tabs until a ticker search exists
```
Comment (iprado78) [inline pages/index.tsx:16]: Handle this with react-rxjs `<Subscribe source={$search} />`

**Review by iprado78 — CHANGES_REQUESTED** [review]

File: components/HistoricalByDay/HistoricalByDay.config.ts [hunk 19]
```diff
@@ -0,0 +1,31 @@
+import { IHistoricalByDay } from "@/models/HistoricalByDay"
+import { toUSD, toUsNum } from "@/utils/numbers"
+
+export const historicalByDayConfig: Record<
+  keyof IHistoricalByDay,
+  {
+    label: string
+    formatter?: (input: any) => any
```
Comment (gracescharf) [inline components/HistoricalByDay/HistoricalByDay.config.ts]:
```suggestion
    formatter?: (input: number) => string
```

File: components/HistoricalByDay/HistoricalByDay.tsx (hunk 2 above)
Comment (gracescharf) [reply components/HistoricalByDay/HistoricalByDay.tsx:2]: * more general to most specific 
* see any of ignacio's prs

File: components/HistoricalByDay/HistoricalByDay.tsx [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+ * - fix z-index issue
+ * - prevent user from selecting a future date
+ * - use date they selected to make an api request
+ *
+ * user searches something. the act of searching creates an observable of their search term.
+ * the data displayed should update if the search term or the date updates.
+ * 1. when the component mounts, get yesterday's data
+ * 2. when the user clicks a button, get the data of the date entered
+ */
+
```
Comment (gracescharf) [inline components/HistoricalByDay/HistoricalByDay.tsx]:
```suggestion

```

File: components/SummaryStats/SummaryStats.tsx [hunk 21]
```diff
@@ -35,40 +26,18 @@ const [useSummaryStats] = bind(
               value: formatter?.(statValue) ?? statValue,
             } as IRow
           }),
-        })),
+        ),
       ),
     ),
   ),
-  {} as SummaryStatsViewModel,
+  {} as IRow[],
```
Comment (iprado78) [inline components/SummaryStats/SummaryStats.tsx]: If you're casting an object as an array, something's gone wrong

File: components/Tabs/Tabs.tsx [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+import styles from "./Tabs.module.scss"
+/**
+ *
+ * @todo implement left-right arrow navigation of tabs:
+ * https://dev.to/eevajonnapanula/keyboard-accessible-tabs-with-react-5ch4
+ */
+
+interface ITabsProps {
+  tabLabels: string[]
+  // is this the right way to type react children?
```
Comment (iprado78) [inline components/Tabs/Tabs.tsx]: Btw, you can put communications to us as self-comments on your PR

File: components/Tabs/Tabs.tsx [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+      aria-selected={isOpen}
+      aria-controls={`tab-panel-${id}`}
+      id={`tab-button-${id}`}
+    >
+      {label}
+    </button>
+  )
+}
+
+const TabPanel = ({ content, isOpen, id }: ITabPanel) => {
```
Comment (iprado78) [inline components/Tabs/Tabs.tsx]: Change content to `content` to `children`, then handle declaratively (see next comment)

File: components/Tabs/Tabs.tsx [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+            isOpen={openTab === label}
+            id={`${label}-${index}`}
+            label={label}
+            key={label}
+          />
+        ))}
+      </div>
+
+      {children.map((child, index) => (
+        <TabPanel
```
Comment (iprado78) [inline components/Tabs/Tabs.tsx:98]:
```
<TabPanel ...>
  { children }
</TabPanel>
```

File: pages/index.tsx (hunk 18 above)
Comment (iprado78) [reply pages/index.tsx:16]: @gracescharf - apologies for sending you down the rabbit hole, but `Subscribe` is likely to not the approach you want.  We can discuss at 9:45

File: components/SummaryStats/SummaryStats.tsx (hunk 21 above)
Comment (gracescharf) [reply components/SummaryStats/SummaryStats.tsx]:
```suggestion
  [] as IRow[],
```

File: components/Tabs/Tabs.module.scss [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+  border: none;
+  font-weight: bold;
+  text-transform: uppercase;
+  line-height: 2;
+  font-size: 1.3rem;
+  color: #0c2074;
+  background: white;
+}
+
+.isActive {
```
Comment (iprado78) [inline components/Tabs/Tabs.module.scss:20]: We should move hex colors into a theme file

**Review by iprado78 — APPROVED** [review]
💯 

@gracescharf - please squash commits then merge

**Pull request merged**